src/cargo/util/context/mod.rs

```diff
@@ -1544,8 +1545,34 @@ impl GlobalContext {
                 // WITH an extension, which people may want to do to
                 // support multiple Cargo versions at once and not
                 // get a warning.
+
+                // On Unix we can see if the two files are the same by using stat(2),
+                // and comparing the inode numbers.  That way it doesn't matter precisely
+                // what spelling the link uses to refers to the other file (relative
+                // vs absolute pathname, etc.)  Also, this is symmetrical: it doesn't mind
+                // which way round the symlink is.
+                #[cfg(unix)]
+                let skip_warning = if let (Ok(metadata_1), Ok(metadata_2)) = (
+                    fs::metadata(&possible),
+                    fs::metadata(&possible_with_extension),
+                ) {
+                    use std::os::unix::fs::MetadataExt;
+                    let devino = |md: fs::Metadata| (md.dev(), md.ino());
+                    devino(metadata_1) == devino(metadata_2)
```

Comment:
Cargo depends on [`same-file`](https://crates.io/crates/same-file) transitively via `cargo-util` btw. We might want to make use of it instead.

Reply:
Interesting.  I was all ready to say "oh I should use that instead".

But I RTFM and `samefile` is unreliable on Windows.  I don't really have an opinion about that (other than "it is impossible to write reliable software on Windows, anyway").  Perhaps it's OK for this to be unreliable on Windows since the only misbehaviour would be to suppress a warning that ought to have been issued.

I would be happy to change this to use `samefile` everywhere.  The documentation is clear about the algorithm on Unix, and it's a reliable one.

LMK what you prefer.